Let RenderAttachment handle the rendering of the selected state

squelart opened the pull request:
#### 0b0286ed2a676c84ff1ef81d26ed2b37e0ce7f84
<pre>
Let RenderAttachment handle the rendering of the selected state
<a href="https://bugs.webkit.org/show_bug.cgi?id=304374">https://bugs.webkit.org/show_bug.cgi?id=304374</a>
<a href="https://rdar.apple.com/111200272">rdar://111200272</a>

Reviewed by Cameron McCormack.

Instead of the default tinted transparent overlay (which modifies all
colors), attachments now handle the selected state by only modifying the
main brick background color, similar to how selected text only changes
its background-color.

When the renderer&apos;s selection state changes, CSS classes are added or
removed to control the eventual painting.

Bonus: The selection state also specifies if the selection ends at the
element or continues outside, so the border radius can be removed when
the selection continues outside of the element, otherwise gaps would be
visible at the corners.

This patch also adds a new type of reference tests, where the actual
attachment brick (with its shadow tree) is compared to a made-up pure
HTML DOM tree with the same styling, but where sub-elements can be
individually tweaked to match test expectations.

* LayoutTests/fast/attachment/cocoa/wide-attachment-ref-progress-075-expected.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-ref-progress-075-selected-alone-expected.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-ref-progress-075-selected-alone.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-ref-progress-075-selected-extended-expected.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-ref-progress-075-selected-extended-left-expected.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-ref-progress-075-selected-extended-left.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-ref-progress-075-selected-extended-right-expected.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-ref-progress-075-selected-extended-right.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-ref-progress-075-selected-extended.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-ref-progress-075.html: Added.
* LayoutTests/fast/attachment/resources/attachment-test-utils.js:
(assertInAttachmentTest): Actually surface script assertion failures
(takeScreenshot):
(takeScreenshotWhenAttachmentsSettled):
(AugmentDivWithAttachmentTree.addChild):
(AugmentDivWithAttachmentTree):
(AugmentAttachmentClassElementsWithAttachmentTree):
* Source/WebCore/html/HTMLAttachmentElement.cpp:
(WebCore::HTMLAttachmentElement::shadowUserAgentStyleSheetText):
(WebCore::HTMLAttachmentElement::ensureWideLayoutShadowTree):
(WebCore::attachmentClassHasSelection):
(WebCore::attachmentClassSelectionContinuesLeft):
(WebCore::attachmentClassSelectionContinuesRight):
(WebCore::HTMLAttachmentElement::addSelectionClasses):
* Source/WebCore/html/HTMLAttachmentElement.h:
* Source/WebCore/html/shadow/attachmentElementShadow.css:
(div#attachment-container.attachment-has-selection):
(div#attachment-container.attachment-selection-continues-left):
(div#attachment-container.attachment-selection-continues-right):
(div#attachment-background):
* Source/WebCore/rendering/RenderAttachment.cpp:
(WebCore::RenderAttachment::setSelectionState):
* Source/WebCore/rendering/RenderAttachment.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::attachmentElementShadowUserAgentStyleSheet const):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Tools/WebKitTestRunner/InjectedBundle/Bindings/TestRunner.idl:
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.h:
(WTR::TestRunner::isVision const):

Canonical link: <a href="https://commits.webkit.org/304918@main">https://commits.webkit.org/304918@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/58a0d4ccec460f4f54142ad18fa464ef34af870a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136866 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/9226 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/48153 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/144602 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/89839 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/623ad3a1-feff-4bf0-a9b6-77de598b5465) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/138738 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9929 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/9072 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104657 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/76384 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/05bbfb11-ffb5-4815-8172-7d6e5f6436f3) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/139811 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/7262 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/122627 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/85495 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/410ca228-894e-46ca-af44-16c263c3c6f0) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6905 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/4597 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/5190 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/116229 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/40810 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/147359 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8909 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/41380 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/113015 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8927 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/7485 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/113344 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28796 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6824 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118910 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/63098 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8957 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36964 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/8678 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/72523 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/8897 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/8749 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

heycam commented on LayoutTests/fast/attachment/resources/attachment-test-utils.js:
```diff
@@ -44,3 +44,344 @@ function takeActualScreenshotWhenAttachmentsSettled() {
+
+    const [platforms, iconSize] = (window.testRunner && window.testRunner.isIOSFamily) ? (window.testRunner.isVision ? [["VISION", "IOS_FAMILY"], 40] : [["IOS_FAMILY"], 72]) : [["MAC"], 52];
+
+    const platformRegex = /#if \(defined\(WTF_PLATFORM_([A-Z_]+)\) && WTF_PLATFORM_\1\)/;
```
Since we hard code the subset of `#define`s that we recognize here, consider failing the test if we find a line starting with `#` that we don't understand.

heycam replied:
An alternative here to embedding a copy of the style sheet may be to use some Internals method to poke into an `<attachment>` element's shadow tree and grab out its style sheet text.

squelart replied:
That shadow tree it not accessible from JS.
Instead I've added an Internals function to retrieve the platform-specific pre-processed style sheet text.

heycam commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.cpp:
```diff
@@ -813,6 +805,14 @@ bool isColorKeywordAllowed(CSSValueID id, const CSSParserContext& context)
+        switch (context.mode) {
+        case UASheetMode:
+        case HTMLQuirksMode:
+            return true;
```
Do we really want to expose these values to quirks mode pages? Is that the only way to have them work in Mail content?

squelart replied:
That's the only way I could find to use these colors in my "expected" test style sheets.
Ideally it'd be nice to have a separate mode for tests.

Otherwise I could replace these color keywords with hard-coded RGBA values when I process the style string, but that could cause troubles if these colors ever change between OS versions!

heycam replied:
It would be better to hard code these, and have someone update the tests if they change, than to expose these non-standard keywords to the web. Or even better might be a way for WKTR to opt in to allowing UA-only values in tests. (Maybe there is already such a thing?)

squelart replied:
Hard-coded values.

heycam commented on LayoutTests/fast/attachment/resources/attachment-test-utils.js:
```diff
@@ -44,3 +61,98 @@ function takeActualScreenshotWhenAttachmentsSettled() {
+    let style = "div.attachment { display: inline-block; margin: 1px; } ";
+    if (window.internals) {
+        style += window.internals.attachmentElementShadowUserAgentStyleSheet();
+
+        style = style.replaceAll("#attachment-", ".attachment-");
+        style = style.replaceAll("-apple-system-tertiary-fill", isIOSFamily ? (isVision ? "rgb(118 118 128 / 0.24)" : "rgb(118 118 128 / 0.12)") : "rgb(0 0 0 / 0.047)");
+        style = style.replaceAll("-apple-system-quaternary-fill", isIOSFamily ? (isVision ? "rgb(118 118 128 / 0.18)" : "rgb(116 116 128 / 0.08)") : "rgb(0 0 0 / 0.03)");
```
Add a comment in here saying that we hard code these values because the `-apple-xxx` color values are only available in UA style sheets.

squelart committed:
Let RenderAttachment handle the rendering of the selected state
https://bugs.webkit.org/show_bug.cgi?id=304374
rdar://111200272

Reviewed by Cameron McCormack.

Instead of the default tinted transparent overlay (which modifies all
colors), attachments now handle the selected state by only modifying the
main brick background color, similar to how selected text only changes
its background-color.

When the renderer's selection state changes, CSS classes are added or
removed to control the eventual painting.

Bonus: The selection state also specifies if the selection ends at the
element or continues outside, so the border radius can be removed when
the selection continues outside of the element, otherwise gaps would be
visible at the corners.

This patch also adds a new type of reference tests, where the actual
attachment brick (with its shadow tree) is compared to a made-up pure
HTML DOM tree with the same styling, but where sub-elements can be
individually tweaked to match test expectations.

* LayoutTests/fast/attachment/cocoa/wide-attachment-ref-progress-075-expected.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-ref-progress-075-selected-alone-expected.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-ref-progress-075-selected-alone.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-ref-progress-075-selected-extended-expected.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-ref-progress-075-selected-extended-left-expected.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-ref-progress-075-selected-extended-left.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-ref-progress-075-selected-extended-right-expected.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-ref-progress-075-selected-extended-right.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-ref-progress-075-selected-extended.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-ref-progress-075.html: Added.
* LayoutTests/fast/attachment/resources/attachment-test-utils.js:
(assertInAttachmentTest): Actually surface script assertion failures
(takeScreenshot):
(takeScreenshotWhenAttachmentsSettled):
(AugmentDivWithAttachmentTree.addChild):
(AugmentDivWithAttachmentTree):
(AugmentAttachmentClassElementsWithAttachmentTree):
* Source/WebCore/html/HTMLAttachmentElement.cpp:
(WebCore::HTMLAttachmentElement::shadowUserAgentStyleSheetText):
(WebCore::HTMLAttachmentElement::ensureWideLayoutShadowTree):
(WebCore::attachmentClassHasSelection):
(WebCore::attachmentClassSelectionContinuesLeft):
(WebCore::attachmentClassSelectionContinuesRight):
(WebCore::HTMLAttachmentElement::addSelectionClasses):
* Source/WebCore/html/HTMLAttachmentElement.h:
* Source/WebCore/html/shadow/attachmentElementShadow.css:
(div#attachment-container.attachment-has-selection):
(div#attachment-container.attachment-selection-continues-left):
(div#attachment-container.attachment-selection-continues-right):
(div#attachment-background):
* Source/WebCore/rendering/RenderAttachment.cpp:
(WebCore::RenderAttachment::setSelectionState):
* Source/WebCore/rendering/RenderAttachment.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::attachmentElementShadowUserAgentStyleSheet const):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Tools/WebKitTestRunner/InjectedBundle/Bindings/TestRunner.idl:
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.h:
(WTR::TestRunner::isVision const):

Canonical link: https://commits.webkit.org/304918@main
